File: echo/server-starnet/Actor/SocketWorker.cs
using System.Collections.Concurrent;
using System.Net.Sockets;

namespace ServerCs.Actor;

public class SocketWorker
{
    private readonly ConcurrentDictionary<int, SocketEvent> _events = new();
    private readonly AutoResetEvent _eventSignal = new(false);
    private volatile bool _running = true;

    public void Init()
    {
        Console.WriteLine("SocketWorker Init");
    }

    public void Stop()
    {
        _running = false;
        _eventSignal.Set(); // Wake up if waiting
    }

    public void Run()
    {
        while (_running)
        {
            ProcessEvents();
            Thread.Sleep(1); // Small delay to prevent CPU spinning
        }
    }

    private void ProcessEvents()
    {
        var socketsToCheck = new List<System.Net.Sockets.Socket>();
        var fdsToCheck = new List<int>();

        foreach (var kvp in _events)
        {
            if (kvp.Value.Socket != null)
            {
                socketsToCheck.Add(kvp.Value.Socket);
                fdsToCheck.Add(kvp.Key);
            }
        }

        if (socketsToCheck.Count == 0)
        {
            return;
        }

        try
        {
            if (socketsToCheck.Count == 0) return;

            var readList = new List<System.Net.Sockets.Socket>(socketsToCheck);
            var writeList = new List<System.Net.Sockets.Socket>();
            var errorList = new List<System.Net.Sockets.Socket>(socketsToCheck);

            System.Net.Sockets.Socket.Select(readList, writeList, errorList, 1); // 1ms timeout

            foreach (var socket in readList)
            {
                int fd = socket.Handle.ToInt32();
                var conn = Starnet.Instance.GetConn(fd);
                if (conn == null) continue;

                if (conn.Type == ConnType.Listen)
                {
                    OnAccept(conn);
                }
                else
                {
                    OnRW(conn, true, false);
                }
            }

            foreach (var socket in writeList)
            {
                int fd = socket.Handle.ToInt32();
                var conn = Starnet.Instance.GetConn(fd);
                if (conn != null && conn.Type == ConnType.Client)
                {
                    OnRW(conn, false, true);
                }
            }
        }
        catch (Exception ex)
        {
            Console.WriteLine($"SocketWorker error: {ex.Message}");
        }
    }

    public void AddEvent(int fd)
    {
        Console.WriteLine($"AddEvent fd {fd}");
        var socket = Starnet.Instance.GetSocket(fd);
        if (socket != null)
        {
            _events[fd] = new SocketEvent { Fd = fd, Socket = socket };
        }
    }

    public void RemoveEvent(int fd)
    {
        Console.WriteLine($"RemoveEvent fd {fd}");
        _events.TryRemove(fd, out _);
    }

    public void ModifyEvent(int fd, bool epollOut)
    {
        Console.WriteLine($"ModifyEvent fd {fd} {epollOut}");
        // In C#, we don't need to modify events like epoll
        // The socket is already set up for both read and write
    }

    private void OnAccept(Conn conn)
    {
        Console.WriteLine($"OnAccept fd: {conn.Fd}");
        var listenSocket = Starnet.Instance.GetSocket(conn.Fd);
        if (listenSocket == null) return;

        try
        {
            var clientSocket = listenSocket.Accept();
            if (clientSocket == null) return;

            clientSocket.Blocking = false;
            clientSocket.NoDelay = true;

            int clientFd = clientSocket.Handle.ToInt32();
            Starnet.Instance.AddConn(clientFd, conn.ServiceId, ConnType.Client);
            Starnet.Instance.RegisterSocket(clientFd, clientSocket);
            AddEvent(clientFd);

            var msg = new SocketAcceptMsg
            {
                Type = MsgType.SocketAccept,
                ListenFd = conn.Fd,
                ClientFd = clientFd
            };
            Starnet.Instance.Send(conn.ServiceId, msg);
        }
        catch (SocketException ex)
        {
            if (ex.SocketErrorCode != SocketError.WouldBlock)
            {
                Console.WriteLine($"OnAccept error: {ex.Message}");
            }
        }
    }

    private void OnRW(Conn conn, bool r, bool w)
    {
        var msg = new SocketRWMsg
        {
            Type = MsgType.SocketRW,
            Fd = conn.Fd,
            IsRead = r,
            IsWrite = w
        };
        Starnet.Instance.Send(conn.ServiceId, msg);
    }

    private class SocketEvent
    {
        public int Fd { get; set; }
        public System.Net.Sockets.Socket? Socket { get; set; }
    }
}

File: echo/server-sunnet/sunnet/src/SocketWorker.cpp
#include "SocketWorker.h"
#include <iostream>
#include <unistd.h>
#include <cassert>
#include <cerrno>
#include <cstring>
#include "Sunnet.h"
#include <sys/socket.h>
#include <fcntl.h>

void SocketWorker::Init()
{
    cout << "SocketWorker Init" << endl;
    epollFd = epoll_create(1024);
    assert(epollFd > 0);
}

void SocketWorker::operator()()
{
    while (true)
    {
        // 阻塞等待
        const int EVENT_SIZE = 64;
        struct epoll_event events[EVENT_SIZE];
        int eventCount = epoll_wait(epollFd, events, EVENT_SIZE, -1);
        // 取得事件
        for (int i = 0; i < eventCount; i++)
        {
            epoll_event ev = events[i]; // 当前要处理的事件
            OnEvent(ev);
        }
    }
}

void SocketWorker::AddEvent(int fd)
{
    cout << "AddEvent fd " << fd << endl;
    // 添加到 epoll 对象
    struct epoll_event ev;
    ev.events = EPOLLIN | EPOLLET;
    ev.data.fd = fd;
    if (epoll_ctl(epollFd, EPOLL_CTL_ADD, fd, &ev) == -1)
    {
        cout << "AddEvent epoll_ctl fail:" << strerror(errno) << endl;
    }
}

void SocketWorker::RemoveEvent(int fd)
{
    cout << "RemoveEvent fd " << fd << endl;
    epoll_ctl(epollFd, EPOLL_CTL_DEL, fd, NULL);
}

void SocketWorker::ModifyEvent(int fd, bool epollOut)
{
    cout << "ModifyEvent fd " << fd << " " << epollOut << endl;
    struct epoll_event ev;
    ev.data.fd = fd;
    if (epollOut)
    {
        ev.events = EPOLLIN | EPOLLET | EPOLLOUT;
    }
    else
    {
        ev.events = EPOLLIN | EPOLLET;
    }
    epoll_ctl(epollFd, EPOLL_CTL_MOD, fd, &ev);
}

void SocketWorker::OnEvent(epoll_event ev)
{
    int fd = ev.data.fd;
    auto conn = Sunnet::inst->GetConn(fd);
    if (conn == NULL)
    {
        cout << "OnEvent error, conn == NULL" << endl;
        return;
    }
    // 事件类型
    bool isRead = ev.events & EPOLLIN;
    bool isWrite = ev.events & EPOLLOUT;
    bool isError = ev.events & EPOLLERR;
    // 监听 socket
    if (conn->type == Conn::TYPE::LISTEN)
    {
        if (isRead)
        {
            OnAccept(conn);
        }
    }
    // 普通 socket
    else
    {
        if (isRead || isWrite)
        {
            OnRW(conn, isRead, isWrite);
        }
        if (isError)
        {
            cout << "OnError fd:" << fd << endl;
        }
    }
}

void SocketWorker::OnAccept(shared_ptr<Conn> conn)
{
    cout << "OnAccept fd:" << conn->fd << endl;
    // 使用边缘触发模式时，需要循环 accept 直到没有更多连接
    while (true)
    {
        // 步骤1：accept
        int clientFd = accept(conn->fd, NULL, NULL);
        if (clientFd < 0)
        {
            if (errno == EAGAIN || errno == EWOULDBLOCK)
            {
                // 没有更多连接可接受
                break;
            }
            cout << "accept error: " << strerror(errno) << endl;
            break;
        }
        // 步骤2：设置非阻塞
        fcntl(clientFd, F_SETFL, O_NONBLOCK);
        // 步骤3：添加连接对象
        Sunnet::inst->AddConn(clientFd, conn->serviceId, Conn::TYPE::CLIENT);
        // 步骤4：添加到 epoll 监听列表
        struct epoll_event ev;
        ev.events = EPOLLIN | EPOLLET;
        ev.data.fd = clientFd;
        if (epoll_ctl(epollFd, EPOLL_CTL_ADD, clientFd, &ev) == -1)
        {
            cout << "OnAccept epoll_ctl fail:" << strerror(errno) << endl;
        }
        // 步骤5：通知服务
        auto msg = make_shared<SocketAcceptMsg>();
        msg->type = BaseMsg::TYPE::SOCKET_ACCEPT;
        msg->listenFd = conn->fd;
        msg->clientFd = clientFd;
        Sunnet::inst->Send(conn->serviceId, msg);
    }
}

void SocketWorker::OnRW(shared_ptr<Conn> conn, bool r, bool w)
{
    auto msg = make_shared<SocketRWMsg>();
    msg->type = BaseMsg::TYPE::SOCKET_RW;
    msg->fd = conn->fd;
    msg->isRead = r;
    msg->isWrite = w;
    Sunnet::inst->Send(conn->serviceId, msg);
}